lib/SILOptimizer/FunctionSignatureTransforms/ExistentialTransform.cpp

```diff
@@ -0,0 +1,590 @@
+    auto *NewGenericParam = GenericTypeParamType::get(Depth, GPIdx++, Ctx);
+    Builder.addGenericParameter(NewGenericParam);
+    Requirement NewRequirement(RequirementKind::Conformance, NewGenericParam,
+                               PType);
+    auto Source =
+        GenericSignatureBuilder::FloatingRequirementSource::forAbstract();
+    Builder.addRequirement(NewRequirement, Source, Mod);
+    ArgToGenericTypeMap.insert(
+        std::pair<int, GenericTypeParamType *>(Idx, NewGenericParam));
+    assert(ArgToGenericTypeMap.find(Idx) != ArgToGenericTypeMap.end());
+  }
+}
+
+/// Create the signature for the newly generated protocol constrained generic
+/// function.
+CanSILFunctionType
+ExistentialTransform::createExistentialSpecializedFunctionType() {
+  auto FTy = F->getLoweredFunctionType();
+  SILModule &M = F->getModule();
+  auto &Ctx = M.getASTContext();
+  GenericSignature *NewGenericSig;
+  GenericEnvironment *NewGenericEnv;
+
+  // Form a new generic signature based on the old one.
+  GenericSignatureBuilder Builder(Ctx);
+
+  /// If the original function is generic, then maintain the same.
+  auto OrigGenericSig = FTy->getGenericSignature();
+  /// First, add the old generic signature.
+  Builder.addGenericSignature(OrigGenericSig);
+
+  /// Convert existential argument types to generic argument types.
+  convertExistentialArgTypesToGenericArgTypes(Builder);
+
+  /// Compute the updated generic signature.
+  NewGenericSig = std::move(Builder).computeGenericSignature(
+      SourceLoc(), /*allowConcreteGenericParams=*/true);
+  NewGenericEnv = NewGenericSig->createGenericEnvironment();
+
+  /// Create a lambda for GenericParams.
+  auto getCanonicalType = [&](Type t) -> CanType {
+    return t->getCanonicalType(NewGenericSig);
+  };
+
+  /// Original list of parameters
+  SmallVector<SILParameterInfo, 4> params;
+  params.append(FTy->getParameters().begin(), FTy->getParameters().end());
+
+  /// Create the complete list of parameters.
+  int Idx = 0;
+  llvm::SmallVector<SILParameterInfo, 8> InterfaceParams;
+  InterfaceParams.reserve(params.size());
+  for (auto &param : params) {
+    auto iter = ArgToGenericTypeMap.find(Idx);
+    if (iter != ArgToGenericTypeMap.end()) {
+      auto GenericParam = iter->second;
+      InterfaceParams.push_back(SILParameterInfo(getCanonicalType(GenericParam),
+                                                 param.getConvention()));
+    } else {
+      auto Ty = param.getType();
+      InterfaceParams.push_back(
+          SILParameterInfo(!Ty->hasTypeParameter()
+                               ? Ty->mapTypeOutOfContext()->getCanonicalType()
+                               : Ty,
+                           param.getConvention()));
+    }
+    Idx++;
+  }
+
+  /// Now add the result types.
+  llvm::SmallVector<SILResultInfo, 8> InterfaceResults;
+  for (auto &result : FTy->getResults()) {
+    auto Ty = result.getType();
+    InterfaceResults.push_back(result.getWithType(
```

Comment:
You can replace this whole thing with InterfaceResults.push_back(result). You don't need to do anything with `Ty`.

Reply:
Or why bother with InterfaceResults at all? You can remove that SmallVector altogether and use FTy->getResults() when you build the new function type instead. The only thing that's changing here is the parameter types, everything else comes directly from the old function type and doesn't need any remapping.